### PR TITLE
server: Ensure batches are either all embed or all completion (#8076)

### LINKED ISSUE #8076: Bug: llama-server crashes when started with --embeddings
### What happened?

System: Mac M1, latest OS (14.5), latest llama.cpp (b3204), build in a std. way (make).
Path to reproduce: start llama server with --embeddings (tested with llama3/8b/fp16 and mistral/7b/Q8_0), go to gui, type anything.
Expected result: system completes/respond.
Actual result: llama-server segfaults: llama_get_logits_ith: invalid logits id 23, reason: no logits / zsh: segmentation fault.
Other notes: The same build/models behave normally when llama-server is started without --embeddings. Similar issue confirmed on Linux/CUDA.

### Name and Version

version: 3204 (45c0e2e4)
built with Apple clang version 15.0.0 (clang-1500.3.9.4) for arm64-apple-darwin23.5.0

### What operating system are you seeing the problem on?

Linux, Mac

### Relevant log output

```shell
[...]
llama_kv_cache_init:      Metal KV buffer size =  4096.00 MiB
llama_new_context_with_model: KV self size  = 4096.00 MiB, K (f16): 2048.00 MiB, V (f16): 2048.00 MiB
llama_new_context_with_model:        CPU  output buffer size =     0.03 MiB
llama_new_context_with_model:      Metal compute buffer size =  2144.00 MiB
llama_new_context_with_model:        CPU compute buffer size =    72.01 MiB
llama_new_context_with_model: graph nodes  = 1030
llama_new_context_with_model: graph splits = 2
INFO [                    init] initializing slots | tid="0x1f31c4c00" timestamp=1719140243 n_slots=1
INFO [                    init] new slot | tid="0x1f31c4c00" timestamp=1719140243 id_slot=0 n_ctx_slot=32768
INFO [                    main] model loaded | tid="0x1f31c4c00" timestamp=1719140243
INFO [                    main] chat template | tid="0x1f31c4c00" timestamp=1719140243 chat_example="[INST] You are a helpful assistant\nHello [/INST]Hi there</s>[INST] How are you? [/INST]" built_in=true
INFO [                    main] HTTP server listening | tid="0x1f31c4c00" timestamp=1719140243 port="8080" n_threads_http="19" hostname="127.0.0.1"
INFO [            update_slots] all slots are idle | tid="0x1f31c4c00" timestamp=1719140243
INFO [      log_server_request] request | tid="0x16b85b000" timestamp=1719140252 remote_addr="127.0.0.1" remote_port=56208 status=200 method="GET" path="/" params={}
INFO [      log_server_request] request | tid="0x16b85b000" timestamp=1719140252 remote_addr="127.0.0.1" remote_port=56208 status=200 method="GET" path="/index.js" params={}
INFO [      log_server_request] request | tid="0x16b8e7000" timestamp=1719140252 remote_addr="127.0.0.1" remote_port=56209 status=200 method="GET" path="/completion.js" params={}
INFO [      log_server_request] request | tid="0x16b973000" timestamp=1719140252 remote_addr="127.0.0.1" remote_port=56211 status=200 method="GET" path="/json-schema-to-grammar.mjs" params={}
INFO [   launch_slot_with_task] slot is processing task | tid="0x1f31c4c00" timestamp=1719140258 id_slot=0 id_task=0
INFO [            update_slots] kv cache rm [p0, end) | tid="0x1f31c4c00" timestamp=1719140258 id_slot=0 id_task=0 p0=0
llama_get_logits_ith: invalid logits id 23, reason: no logits
zsh: segmentation fault
```

### PR DESCRIPTION
**Pull request by iamlemec**
fix #8076 

This ensures that batches are never mixed between embedding and completion tasks. It also adds a check in completion handlers ensuring the server was not run in embedding mode (otherwise it would crash for embedding-only models). Thus to allow for completions and embeddings on the same server, one must run it *without* the `--embeddings` flag.

Currently, this does not work well with `n_parallel > 1` since we are changing the `embedding` flag on a shared context, which will lead to missing logits and a failed assertion in many cases. Not sure if that's a deal-breaker.

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [X] Low
  - [ ] Medium
  - [ ] High

### CODE REVIEW

File: examples/server/server.cpp [hunk 1]
```diff
@@ -2173,6 +2177,14 @@ struct server_context {
                         }
                     }
 
+                    // check that we are in the right batch_type, if not defer the slot
+                    bool slot_type = slot.embedding ? 1 : 0;
+                    if (batch_type == -1) {
+                        batch_type = slot_type;
+                    } else if (batch_type != slot_type) {
+                        continue;
+                    }
```
Comment (compilade) [inline examples/server/server.cpp:2186]: Should this be done earlier? There's some things like `slot.n_past` which should probably not be changed when the batch_type is not the right one.

I feel like this check should be done at least before tokens are put in the `batch`, but the multitude of loops over slots does make it hard to find one right spot.

Comment (iamlemec) [reply examples/server/server.cpp:2186]: Indeed. But we also want to make sure we don't assign the batch to a particular type but then bail out later on (say, because the prompt is bigger than the batch size).

My current read is that everything before where the check is currently is tokenization, and it's fine to do that, bail, and then pick up the slot on the next go around. That's why it's okay to have the conditional `continue` just above it (2173-2178) too.

Comment (compilade) [reply examples/server/server.cpp:2186]: > everything before where the check is currently is tokenization and it's fine to do that, bail, and then pick up the slot on the next go around.

Right, and `slot.n_past` is recalculated anyway, so you're right.

But I think if `batch.n_tokens` is non-zero, the initial `batch_type` should be `non-embedding`, or else this could potentially attempt continuous batching with embeddings *and* previous text completions.

Comment (iamlemec) [reply examples/server/server.cpp:2187]: Ah I see, just pushed the fix!

**Comment by ggerganov** [comment]
> Currently, this does not work well with n_parallel > 1 since we are changing the embedding flag on a shared context, which will lead to missing logits and a failed assertion in many cases

I don't see why it wouldn't work. `n_parallel > 1` means more slots and this new logic with the batch type now correctly handles multiple slots. Can you provide a fail case?

**Comment by iamlemec** [comment]
> I don't see why it wouldn't work. `n_parallel > 1` means more slots and this new logic with the batch type now correctly handles multiple slots. Can you provide a fail case?

You know, it was crashing before when I tried requesting a completion and an embedding at the same time (with `-np 2`), but now it's running fine. It's possible the initial `batch_type` fix discussed above was what was happening.

**Review by ggerganov — APPROVED** [review]

**Comment by okigan** [comment]
It appears that with this change, either the completion endpoint or the embeddings endpoint will work, but not both simultaneously. Does this represent a loss of functionality, or did it simply not function correctly before, without proper error handling?

**Comment by iamlemec** [comment]
@okigan If you're using a generative (non-embedding) model, you can run it as usual, and it will be able to serve both types of endpoints. In that case, you'll probably want to provide it with a `--pooling` option to tell it how to make sequence embeddings. If you're using an embedding-only model, you need to pass `--embeddings` to the server when you launch it, and it will only be able to provide embeddings (by design).

**Comment by okigan** [comment]
@iamlemec so seems like "--embeddings" flag shall be used with "embedding" only model, right? how would one identify if model file is for embeddings only? and if embeddings only model is passed would llama-server require the flag?

I also made a [script](https://github.com/okigan/llama-cpp-scripts/blob/main/scripts/test_llama_server.sh) to test this with mistral model (3rd response fails with 501):

    ./test_llama_server.sh
    Model already exists. Skipping download.
    Running server WITHOUT --embeddings flag...
    Waiting for server to start...
    Server is ready.
    Testing chat completion...
    > POST /v1/chat/completions HTTP/1.1
    < HTTP/1.1 200 OK
    
    
    Testing embeddings...
    > POST /v1/embeddings HTTP/1.1
    < HTTP/1.1 200 OK
    
    #################
    
    Running server with --embeddings flag...
    Waiting for server to start...
    Server is ready.
    Testing chat completion...
    > POST /v1/chat/completions HTTP/1.1
    < HTTP/1.1 501 Not Implemented                 <<<<<---- here
    
    
    Testing embeddings...
    > POST /v1/embeddings HTTP/1.1
    < HTTP/1.1 200 OK


so, if the above expected - i guess the docs are confusing for the --embeddings flag, would you agree?

**Comment by iamlemec** [comment]
@okigan Yeah, `--embeddings` would make most sense for embedding-only models, or perhaps if you want to ensure that completions are not served.

Funily enough, despite the many configuation flags seen in HF world, there isn't one that cleanly identifies whether it's an embedding model or not. Usually embedding models are clearly identified as such in the README. If you want to do it programatically, the best way right now is to look at the `hparams.causal_attn` flag. There are probably some exceptions, but embedding models typically have `causal_attn = False` while generative models have `causal_attn = True`. I think there was talk at some point of auto-detecting based on this heuristic, but it hasn't been done.

The main reason this slightly awkward setup is needed is that if you try to get a completion from an embedding model it'll actually crash. Basically, it looks for logits in the output, doesn't find them, and gives up. So you want to protect embedding models from crashing themselves if you ask for a completion.

I agree, the docs are not really accurate anymore (at least what's written in the command line help). It should probably now say something like "embedding only mode" rather than "enable embedding endpoint".

**Comment by okigan** [comment]
@iamlemec @ggerganov trying to clarify this, created https://github.com/ggerganov/llama.cpp/issues/8763, please review and comment there. If we have a clear path forward I can propose a PR.
